### PR TITLE
Two implementations of Grid-based layout engines

### PR DESCRIPTION
**Pull request by N1x0**
This is to address issue #136 by adding two new `LayoutEngines`. One is a *pseudo-grid* with a fixed number of rows (2),while the other scales dynamically i.e. it adjusts the number of rows as well.

I have opted to keep my first solution (pseudo grid) because it looks nicer for a certain number of windows (5, 7,11) and I couldn't figure out a neat way to structure the "true" Grid the same way.

As stated in the original issue none of the (re)-sizing functions are implemented so what I'd suggest is to load the workspace with the `FullLayoutEngine` as secondary layout
*e.g.*
`sticky.CreateWorkspace("GRID", new GridLayoutEngine(), new FullLayoutEngine());`

Open to any comments and suggestions 😄

### CODE REVIEW

**Comment by josteink** [comment]
I'm no  master of custom layouts honestly, but I did skim over the code and it looks good to me.

If I would have any suggestions it would be to use less "clever" and inlined code for calculating the number of rows and columns. They are kinda hard to read as is.

**Comment by N1x0** [comment]
> 
> 
> I'm no master of custom layouts honestly, but I did skim over the code and it looks good to me.
> 
> If I would have any suggestions it would be to use less "clever" and inlined code for calculating the number of rows and columns. They are kinda hard to read as is.

Thanks for your quick reply @josteink , I adjusted the ternary  to be a proper `if-else` block for the pseudo grid and rewrote `Math.Pow(numWindows, 0.5)` to `Math.Sqrt(numWindows)` for the true grid which should make it a bit more readable.

Happy if @BeeryShklar could give them a spin too since he originally made the request for this addition.

File: src/workspacer.Shared/Layout/GridLayoutEngine.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      
+
+                int rows = (int)(Math.Sqrt(numWindows) + 0.5);
+
+                while (numWindows % rows != 0)
+                {
+                    rows -= 1;
+                }
+
+                int cols = numWindows / rows;
```
Comment (josteink) [inline src/workspacer.Shared/Layout/GridLayoutEngine.cs:29]: These lines seems to be indented 4 characters too much, making them appear as if they are part of the if-block.

Comment (N1x0) [reply src/workspacer.Shared/Layout/GridLayoutEngine.cs]: my bad, fixed in newest commit

**Comment by josteink** [comment]
LGTM with latest changes. :+1:

Any objections from anyone else?

**Comment by BeeryShklar** [comment]
@N1x0  sorry for the late response, I tried running your fork and couldn't.
This is what I have done:
I added your fork as a remote, made a new branch and cloned your `GridLayouts` branch.
When I ran `dotnet run` from `src/workspacer` it failed.
This is the log:
```
D:\code\rickbutton\workspacer\src\workspacer\workspacer.csproj(59,5): error MSB3030: Could not copy the file "D:\code\rickbutton\workspacer\\src\workspacer.Bar\bin\Debug\netcoreapp3.1\win-x64\workspacer.Bar.dll" because it was not found.
```
Can someone help me run it? I am *very* new to dotnet and workspacer is the first time i have actually coded in c#.

**Comment by N1x0** [comment]
Hi @BeeryShklar  thanks for wanting to try it out. So the error message is pretty clear I think, no? Doesn't seem like the `.dll` file was found. I tried the same for me and it worked, but also my `workspacer.bar.dll` was in a separate folder. I'd recommend just using visual studio instead for debugging and testing via `dotnet run`

**Comment by rickbutton** [comment]
Thanks for the contribution! I'd love to have more grid-based layouts available.

I think that we may be able to utilize the increment/decrement methods in order to determine the number of rows in the grid. Does anyone like that idea?

**Comment by josteink** [comment]
That sounds like a good idea to me. I'm all for it.

**Review by josteink — APPROVED** [review]
LGTM. We can eventually add the resize => row functionionality later on, if people ask for it.

**Comment by josteink** [comment]
If @rickbutton has approved, there's really no good reason not to merge this.

This has what... Been hanging around for 3-4 months now? :fearful: 

Let's get it done!

**Comment by josteink** [comment]
If rickbutton also likes this, there's really no good reason not to merge this. This has what... Been hanging around for 3-4 months now? :fearful: 

But it seems like at least 2 reviews are required now. Holding on until @rickbutton or @dalyIsaac get's around to approve.

**Comment by dalyIsaac** [comment]
I quite like the pseudo-grid. I do find some of the transitions to columns in the dynamically-scaling grid are slightly disconcerting, which I think @N1x0 was referring to.

The logic probably could be updated in the future, particularly with the resize => row functionality.

However, **it's an awesome start.**

LGTM

**Review by dalyIsaac — APPROVED** [review]

**Comment by josteink** [comment]
"At least 2 approving reviews are required by reviewers with write access."

Oh well. Guess it's just me and @rickbutton then?

**Pull request closed without merging**